lib/req/steps.ex

```diff
@@ -330,12 +330,16 @@ defmodule Req.Steps do
   end
 
   defp put_params(request, params) do
-    encoded = URI.encode_query(params)
+    if Req.Request.get_private(request, :req_retry_count) do
+      request
```

Comment:
This is a pretty sensible way to fix the issue but I can't help but wonder if there's another way. There very well may not be a better option. But it means that any step that changes the request somehow, which is most of them, would need to have similar conditional and it is weird to hardcode it for retry.

Another example where this is clearly broken is compress_body/1 step, on retries we'll keep re-compressing the request body, getting a different result each time.

I think we have these options:

1. This PR
2. Require steps to be idempotent. For example, put_params/1 could check if given params are already in the query string. (Which would prevent us from sending duplicate ones.) compress_body/1 could check if content-encoding is already set and if so, it's a no-op, etc.
3. When a step is executed, remove the option that triggered it. This seems fragile.
4. When a step is executed, remove it from the steps list. I vaguely remember we had plugins which checked if a given request contains given steps, so this would break that.
5. Option 4 except we keep two lists, the latter being private and the one that gets stuff deleted from.

Any other ideas?

Reply:
4/5. when we run a step that returns on error, it is executed, but at the same time, for retry we would like to repeat that step, how do we distinguish that?

One other option would be to store the initial `%Req.Request{}` in a separate field when calling `run/1`, then if we need to re-run we just use that intact request, only bumping the private field. Though with this approach we redo stuff like compression.